File: shared/validators/validate_message.py
"""IPC message validation helpers for backend services."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID


REQUIRED_KEYS = {"type", "payload"}
OPTIONAL_KEYS = {"timestamp", "request_id"}


def _is_iso8601(value: str) -> bool:
    try:
        datetime.fromisoformat(value.replace("Z", "+00:00"))
        return True
    except ValueError:
        return False


def _is_uuid(value: str) -> bool:
    try:
        UUID(value)
        return True
    except ValueError:
        return False


def validate_message_envelope(message: Any) -> tuple[bool, str | None]:
    """Validate common IPC envelope fields.

    Returns:
        (True, None) on success, (False, reason) on validation failure.
    """
    if not isinstance(message, dict):
        return False, "Message must be an object"

    missing = REQUIRED_KEYS - set(message.keys())
    if missing:
        return False, f"Missing required keys: {', '.join(sorted(missing))}"

    unknown = set(message.keys()) - REQUIRED_KEYS - OPTIONAL_KEYS
    if unknown:
        return False, f"Unknown envelope keys: {', '.join(sorted(unknown))}"

    if not isinstance(message["type"], str) or not message["type"].strip():
        return False, "Field 'type' must be a non-empty string"

    if not isinstance(message["payload"], dict):
        return False, "Field 'payload' must be an object"

    if "timestamp" in message and not _is_iso8601(str(message["timestamp"])):
        return False, "Field 'timestamp' must be ISO-8601"

    if "request_id" in message and not _is_uuid(str(message["request_id"])):
        return False, "Field 'request_id' must be a UUID"

    return True, None

File: shared/validators/validate_message.ts
export type IpcEnvelope = {
  type: string;
  payload: Record<string, unknown>;
  timestamp?: string;
  request_id?: string;
};

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

const isIsoDate = (value: string): boolean => {
  const date = new Date(value);
  return !Number.isNaN(date.getTime());
};

export function validateMessageEnvelope(
  message: unknown,
): { valid: boolean; error?: string } {
  if (typeof message !== "object" || message === null || Array.isArray(message)) {
    return { valid: false, error: "Message must be an object" };
  }

  const obj = message as Record<string, unknown>;

  if (typeof obj.type !== "string" || !obj.type.trim()) {
    return { valid: false, error: "Field 'type' must be a non-empty string" };
  }

  if (
    typeof obj.payload !== "object" ||
    obj.payload === null ||
    Array.isArray(obj.payload)
  ) {
    return { valid: false, error: "Field 'payload' must be an object" };
  }

  if (obj.timestamp !== undefined) {
    if (typeof obj.timestamp !== "string" || !isIsoDate(obj.timestamp)) {
      return { valid: false, error: "Field 'timestamp' must be ISO-8601" };
    }
  }

  if (obj.request_id !== undefined) {
    if (typeof obj.request_id !== "string" || !UUID_RE.test(obj.request_id)) {
      return { valid: false, error: "Field 'request_id' must be a UUID" };
    }
  }

  return { valid: true };
}
